salt/_modules/metalk8s_volumes.py

```diff
@@ -0,0 +1,430 @@
+    """
+    return _get_volume(name).finalize()
+
+
+def is_formatted(name):
+    """Check if the given volume is formatted.
+
+    Args:
+        name (str): volume name
+
+    Returns:
+        bool: True if the volume is already formatted, otherwise False
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' metalk8s_volumes.volume_is_formatted example-volume
+    """
+    return _get_volume(name).is_formatted
+
+
+def format(name):
+    """Format the given volume.
+
+    Args:
+        name (str): volume name
+
+    Returns:
+        None
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' metalk8s_volumes.format example-volume
+    """
+    _get_volume(name).format()
+
+
+# Volume {{{
+
+
+class Volume(object):
+    """Volume interface."""
+
+    __metaclass__ = abc.ABCMeta
+
+    def __init__(self, volume):
+        self._volume = volume
+
+    @abc.abstractproperty
+    def exists(self):
+        """Does the backing storage device exists?"""
+        return
+
+    @abc.abstractmethod
+    def create(self):
+        """Create the backing storage device."""
+        return
+
+    @abc.abstractproperty
+    def is_finalized(self):
+        """Check if the backing storage device setup is finalized."""
+        return
+
+    @abc.abstractmethod
+    def finalize(self):
+        """Finalize the setup of the backing storage device."""
+        return
+
+    @abc.abstractproperty
+    def path(self):
+        """Path to the backing device."""
+        return
+
+    @property
+    def is_formatted(self):
+        """Check if the volume is already formatted."""
+        fs_type = self.get('spec.storageClassName.parameters.fsType')
+        return _get_fstype(self.path) == fs_type
+
+    def format(self, force=False):
+        """Format the volume.
+
+        The volume is formatted according to its StorageClass.
+        """
+        # Check that the backing device is not already formatted.
+        # Bail out if it is: we don't want data loss because of a typo…
+        if _get_fstype(self.path):
+            raise Exception('backing device `{}` already formatted'.format(
+                self.path
+            ))
+        params = self.get('spec.storageClassName.parameters')
+        # mkfs options, if any, are stored as JSON-encoded list.
+        options = json.loads(params.get('mkfsOptions', '[]'))
+        fs_type = params['fsType']
+        command = _mkfs(
+            self.path, fs_type, self.get('metadata.uid'), force, options
+        )
+        _run_cmd(' '.join(command))
+
+    def get(self, path):
+        """Return the Volume attribute `path` from the Volume dict."""
+        return functools.reduce(operator.getitem, path.split('.'), self._volume)
+
+
+# }}}
+# SparseLoopDevice {{{
+
+
+class SparseLoopDevice(Volume):
+    @property
+    def path(self):
+        return '/var/lib/metalk8s/storage/sparse/{}'.format(
+            self.get('metadata.name')
```

Comment:
I'm not entirely sure using `name` is the right approach, I'd argue UUID would be better.

Rationale: one can create a volume with some name, then things can go wrong, one could then (possibly much later) create another volume with the same name, and get a conflict.

Reply:
Yeah UUID is better, the PR started with `name` and I kept it but as you said: UUID is safer;
I'll make the change.

Reply:
Done